File: clients/android/NewsBlur/src/com/newsblur/fragment/ProfileDetailsFragment.java
package com.newsblur.fragment;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.os.Bundle;
import android.app.Fragment;
import android.app.FragmentManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.newsblur.R;
import com.newsblur.domain.UserDetails;
import com.newsblur.network.APIManager;
import com.newsblur.util.FeedUtils;
import com.newsblur.util.PrefsUtils;
import com.newsblur.util.UIUtils;

public class ProfileDetailsFragment extends Fragment implements OnClickListener {
	
	UserDetails user;
	private TextView username, bio, location, sharedCount, followerCount, followingCount, website;
	private View locationIcon;
	private ImageView imageView;
	private boolean viewingSelf = false;
	private Button followButton, unfollowButton;
	private APIManager apiManager;
	
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		apiManager = new APIManager(getActivity());
	}
	
	public void setUser(Context context, final UserDetails user, final boolean viewingSelf) {
		this.user = user;
		this.viewingSelf = viewingSelf;
		if (username != null) {
			setUserFields(context);
		}
	}
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View v = inflater.inflate(R.layout.fragment_profiledetails, null);
		username = (TextView) v.findViewById(R.id.profile_username);
		bio = (TextView) v.findViewById(R.id.profile_bio);
		location = (TextView) v.findViewById(R.id.profile_location);
		locationIcon = v.findViewById(R.id.profile_location_icon);
		sharedCount = (TextView) v.findViewById(R.id.profile_sharedcount);
		website = (TextView) v.findViewById(R.id.profile_website);
		followerCount = (TextView) v.findViewById(R.id.profile_followercount);
		followingCount = (TextView) v.findViewById(R.id.profile_followingcount);
		imageView = (ImageView) v.findViewById(R.id.profile_picture);
		followButton = (Button) v.findViewById(R.id.profile_follow_button);
		unfollowButton = (Button) v.findViewById(R.id.profile_unfollow_button);
		followButton.setOnClickListener(this);
		unfollowButton.setOnClickListener(this);
		
		if (user != null) {
			setUserFields(getActivity());
		}
		
		return v;
	}

	private void setUserFields(Context context) {
		username.setText(user.username);
		
		if (!TextUtils.isEmpty(user.bio)) {
			bio.setText(user.bio);
		} else {
			bio.setVisibility(View.INVISIBLE);
		}
		
		if (!TextUtils.isEmpty(user.location)) {
			location.setText(user.location);
		} else {
			location.setVisibility(View.INVISIBLE);
			locationIcon.setVisibility(View.INVISIBLE);
		}
		
		if (!TextUtils.isEmpty(user.website)) {
			website.setText("" + user.website);
		} else {
			website.setVisibility(View.GONE);
		}
		
		sharedCount.setText("" + user.sharedStoriesCount);
		
		followerCount.setText("" + user.followerCount);
		
		followingCount.setText("" + user.followingCount);
		
		if (!viewingSelf) {
			FeedUtils.iconLoader.displayImage(user.photoUrl, imageView, 5, false);
			if (user.followedByYou) {
				unfollowButton.setVisibility(View.VISIBLE);
				followButton.setVisibility(View.GONE);
			} else {
				unfollowButton.setVisibility(View.GONE);
				followButton.setVisibility(View.VISIBLE);
			}
		} else {
			followButton.setVisibility(View.GONE);
			Bitmap userPicture = PrefsUtils.getUserImage(context);
			// seems to sometimes be an error loading the picture so prevent
			// force close if null returned
			if (userPicture != null) {
		        userPicture = UIUtils.clipAndRound(userPicture, 5, false);
			    imageView.setImageBitmap(userPicture);
			}
		}
	}
	
	private class FollowTask extends AsyncTask<Void, Void, Boolean> {
		@Override
		protected void onPreExecute() {
			followButton.setEnabled(false);
		}
		
		@Override
		protected Boolean doInBackground(Void... params) {
			return apiManager.followUser(user.userId);
		}
		
		@Override
		protected void onPostExecute(Boolean result) {
			followButton.setEnabled(true);
			if (result) {
				user.followedByYou = true;
				followButton.setVisibility(View.GONE);
				unfollowButton.setVisibility(View.VISIBLE);
			} else {
				FragmentManager fm = ProfileDetailsFragment.this.getFragmentManager();
		        AlertDialogFragment alertDialog = AlertDialogFragment.newAlertDialogFragment(getResources().getString(R.string.follow_error));
		        alertDialog.show(fm, "fragment_edit_name");
			}	
		}		
	}
	
	private class UnfollowTask extends AsyncTask<Void, Void, Boolean> {
		@Override
		protected void onPreExecute() {
			unfollowButton.setEnabled(false);
		}
		
		@Override
		protected Boolean doInBackground(Void... params) {
			return apiManager.unfollowUser(user.userId);
		}
		
		@Override
		protected void onPostExecute(Boolean result) {
			unfollowButton.setEnabled(true);
			if (result) {
				user.followedByYou = false;
				unfollowButton.setVisibility(View.GONE);
				followButton.setVisibility(View.VISIBLE);
			} else {
				FragmentManager fm = ProfileDetailsFragment.this.getFragmentManager();
                AlertDialogFragment alertDialog = AlertDialogFragment.newAlertDialogFragment(getResources().getString(R.string.unfollow_error));
                alertDialog.show(fm, "fragment_edit_name");
			}	
		}		
	}
	

	@Override
	public void onClick(View v) {
		if (v.getId() == R.id.profile_follow_button) {
			new FollowTask().execute();
		} else if (v.getId() == R.id.profile_unfollow_button) {
			new UnfollowTask().execute();
		}
	}
	
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/fragment/ProfileDetailsFragment.kt
package com.newsblur.fragment

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.newsblur.R
import com.newsblur.databinding.FragmentProfiledetailsBinding
import com.newsblur.di.IconLoader
import com.newsblur.domain.UserDetails
import com.newsblur.network.APIManager
import com.newsblur.preference.PrefRepository
import com.newsblur.util.ImageLoader
import com.newsblur.util.UIUtils
import com.newsblur.util.executeAsyncTask
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

@AndroidEntryPoint
class ProfileDetailsFragment : Fragment() {

    @Inject
    lateinit var apiManager: APIManager

    @IconLoader
    @Inject
    lateinit var iconLoader: ImageLoader

    @Inject
    lateinit var prefRepository: PrefRepository

    private var user: UserDetails? = null
    private var viewingSelf = false

    private lateinit var binding: FragmentProfiledetailsBinding

    fun setUser(user: UserDetails?, viewingSelf: Boolean) {
        this.user = user
        this.viewingSelf = viewingSelf
        if (::binding.isInitialized) {
            setUserFields()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_profiledetails, container, false)
        binding = FragmentProfiledetailsBinding.bind(view)
        binding.profileFollowButton.setOnClickListener { followUser() }
        binding.profileUnfollowButton.setOnClickListener { unfollowUser() }
        user?.let { setUserFields() }
        return view
    }

    private fun setUserFields() {
        binding.profileUsername.text = user!!.username
        if (!TextUtils.isEmpty(user!!.bio)) {
            binding.profileBio.text = user!!.bio
        } else {
            binding.profileBio.visibility = View.INVISIBLE
        }
        if (!TextUtils.isEmpty(user!!.location)) {
            binding.profileLocation.text = user!!.location
        } else {
            binding.profileLocation.visibility = View.INVISIBLE
            binding.profileLocationIcon.visibility = View.INVISIBLE
        }
        if (!TextUtils.isEmpty(user!!.website)) {
            binding.profileWebsite.text = user!!.website
        } else {
            binding.profileWebsite.visibility = View.GONE
        }
        binding.profileUserStatistics.profileSharedcount.text = user!!.sharedStoriesCount.toString()
        binding.profileUserStatistics.profileFollowercount.text = user!!.followerCount.toString()
        binding.profileUserStatistics.profileFollowingcount.text = user!!.followingCount.toString()
        if (!viewingSelf) {
            iconLoader.displayImage(user!!.photoUrl, binding.profilePicture)
            if (user!!.followedByYou) {
                binding.profileUnfollowButton.visibility = View.VISIBLE
                binding.profileFollowButton.visibility = View.GONE
            } else {
                binding.profileUnfollowButton.visibility = View.GONE
                binding.profileFollowButton.visibility = View.VISIBLE
            }
        } else {
            binding.profileFollowButton.visibility = View.GONE
            setUserPicture()

        }
    }

    private fun setUserPicture() {
        lifecycleScope.launch(Dispatchers.IO) {
            val userPicture = prefRepository.getUserImage(requireContext())
            if (userPicture != null) {
                val roundedPicture = UIUtils.clipAndRound(userPicture, true, false)
                withContext(Dispatchers.Main) {
                    binding.profilePicture.setImageBitmap(roundedPicture)
                }
            }
        }
    }

    private fun followUser() {
        lifecycleScope.executeAsyncTask(
                onPreExecute = {
                    binding.profileFollowButton.isEnabled = false
                },
                doInBackground = {
                    apiManager.followUser(user!!.userId)
                },
                onPostExecute = {
                    binding.profileFollowButton.isEnabled = true
                    if (it) {
                        user!!.followedByYou = true
                        binding.profileFollowButton.visibility = View.GONE
                        binding.profileUnfollowButton.visibility = View.VISIBLE
                    } else {
                        val alertDialog = AlertDialogFragment.newAlertDialogFragment(resources.getString(R.string.follow_error))
                        alertDialog.show(parentFragmentManager, "fragment_edit_name")
                    }
                }
        )
    }

    private fun unfollowUser() {
        lifecycleScope.executeAsyncTask(
                onPreExecute = {
                    binding.profileUnfollowButton.isEnabled = false
                },
                doInBackground = {
                    apiManager.unfollowUser(user!!.userId)
                },
                onPostExecute = {
                    binding.profileUnfollowButton.isEnabled = true
                    if (it) {
                        user!!.followedByYou = false
                        binding.profileUnfollowButton.visibility = View.GONE
                        binding.profileFollowButton.visibility = View.VISIBLE
                    } else {
                        val alertDialog = AlertDialogFragment.newAlertDialogFragment(resources.getString(R.string.unfollow_error))
                        alertDialog.show(parentFragmentManager, "fragment_edit_name")
                    }
                }
        )
    }
}